### DISCUSSION TITLE
[@types/wicg-file-system-access] not compatible with `lib.dom.d.ts`

### DISCUSSION THREAD

**Discussion by jespertheend** [DefinitelyTyped/DefinitelyTyped#58386] [Issues with a @types package]
I'm using the latest dev version of TypeScript and this seems to have partially included the File System Access api in `lib.dom.d.ts`. My guess is that it's because [safari supports it now](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle#browser_compatibility). But some methods like [`queryPermission`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle/queryPermission) and [`requestPermission`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemHandle/requestPermission) are not implemented and so they're not included in `lib.dom.d.ts`.

The problem is, even though I've installed `@types/wicg-file-system-access`, these methods are still not available. If I use TypeScript `4.5.5` it works fine, but with `4.6.0-dev.20220123` it doesn't.

@RReverser since you wrote the original types, do you have any idea how this could be fixed? I think the problem is that `wicg-file-system-access` uses classes for `FileSystemFileHandle` and `FileSystemDirectoryHandle`, whereas `lib.dom.d.ts` uses interfaces.

[a comment by an automated account is not shown]

**Comment by RReverser** [comment]
Hmm I'd expect types to get merged together - afaik that works fine even for classes+interfaces in TypeScript.

Do you see any specific errors? Unfortunately, I can't dig into it deeper right now but if you find the culprit, I'll be happy to help with reviews and/or ideas for fixes.

**Reply by jespertheend** [reply]
Sure, I’ll try to create a reduced test case.

**Reply by jespertheend** [reply]
Hmm it seems there's a bunch more issues I hadn't noticed before. But it's mostly just a bunch of `TS2300 Duplicate identifier` errors. I tried changing the three classes to an interface and that seems to work for most of the errors.

Only the duplicate `type FileSystemHandleKind = 'file' | 'directory'` I'm not sure how to fix.

In case you wanted to try it out, this the reduced test case I'm using: [fsaTypes.zip](https://github.com/DefinitelyTyped/DefinitelyTyped/files/7921215/fsaTypes.zip)
To run it simply doing `npm i` and `npm run it` should do it. That should run tsc `4.6.0-dev.20220123` and output all errors.
If you open it in VSCode and accept the typescript version dialog I think the errors should show up in your ide as well.

I'll submit a PR with the things I managed to fix so far.

**Reply by jespertheend** [reply]
I've submitted https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58389. But as expected, ci is failing for the reasons I described earlier.

**Comment by bogacg** [comment]
Here is an error I see:
```
Property 'createSyncAccessHandle' does not exist on type 'FileSystemFileHandle'.
```
I guess that's a relatively new method.

### Update
Further digging, apparently 2022 was an important year for File System Access API. Discussions held, updates were made.
* [FileSystemSyncAccessHandle contains async methods · Issue #7 · whatwg/fs](https://github.com/whatwg/fs/issues/7)
* [Breaking change: sync methods for AccessHandles - Chrome Developers](https://developer.chrome.com/blog/sync-methods-for-accesshandles/)

**Comment by masylum** [comment]
Same error here, looks like there are no definitions for `createSyncAccessHandle`
